### ISSUE TITLE
12.4 深度学习之CIFAR-10二进制数据集

### ISSUE THREAD

**Issue by vieyahn2017** [vieyahn2017/iBlog#333]
cifar10

**Comment by vieyahn2017** [comment]
## TensorFlow实现用于图像分类的卷积神经网络（代码详细注释）
https://blog.csdn.net/qrlhl/article/details/78608985


这里我们采用cifar10作为我们的实验数据库。
首先下载TensorFlow Models库，以便使用其中提供的CIFAR-10数据的类。

git clone https://github.com/tensorflow/models.git
cd models/tutorials/image/cifar10

**Comment by vieyahn2017** [comment]
## 使用Tensorflow读取CIFAR-10二进制数据集
https://www.cnblogs.com/cloud-ken/p/8458192.html

参考文献
Tensorflow官方文档
tf.transpose函数解析
tf.slice函数解析
CIFAR10/CIFAR100数据集介绍
tf.train.shuffle_batch函数解析
Python urllib urlretrieve函数解析

**Comment by vieyahn2017** [comment]
## 从cifar10分类入门深度学习图像分类（Keras）

引
之前需要做一个图像分类模型，因为刚入门，拿cifar10数据集练了下手，试了几种优化方案和不同的模型效果，这里就统一总结一下这段学习经历。

对于新手来说，最方便的深度学习框架应该是Keras了，这是一个可以基于Tensorflow、PyTorch等多种深度学习框架的高级框架，用它来搭建和训练模型特别方便，很适合入门快速掌握。这里推荐下Keras之父写的深度学习入门书籍：《Python深度学习》，轻松易懂，而且直接教你一步步用Keras搭建模型，比看西瓜书《机器学习》和花书《深度学习》效率快很多。

cifar10是是一个图像数据集（官网），包含10种类别的32*32大小的图像共60000张。另外还有cifar100，包含100种类别的更多图像。因此，cifar10分类就是一个图像多分类任务。Keras另一个好处在于已经集成了很多常见的数据集和模型，在接口里可以直接调用，当然，为了减小安装包，会在你第一次调用的时候才进行下载，但因为某些因素可能直接下载会失败，因此也可以自己先下载好后再使用，至于下载的方法网上有很多。

因此，本文要说的就是使用Keras框架来开发多种模型和优化方法去训练一个基于cifar10数据集的图像多分类模型。

环境：
本文代码运行的环境为Linux，Tensorflow 1.4，Keras 2.1.1，其他的库版本就不说啦，直接安装即可，当然，跑训练最重要的还是最好有一个性能强劲的GPU。

目录
简单CNN训练
简单CNN数据增强
基于VGG16-imagenet进行特征提取
基于VGG16-imagenet进行模型微调（fine-tuning）
深层CNN训练
简单Resnet训练

整体工程：https://github.com/Cloudox/Cifar10-Classify

**Comment by vieyahn2017** [comment]
用python将cifar-10二进制数据集转成图片格式
https://blog.csdn.net/guohuifengby/article/details/62424299

```python

#encoding:utf-8
#author：guohuifeng
from scipy.misc import imsave
import numpy as np

# 解压缩，返回解压后的字典
def unpickle(file):
    import cPickle
    fo = open(file, 'rb')
    dict = cPickle.load(fo)
    fo.close()
    return dict

# 生成训练集图片，如果需要png格式，只需要改图片后缀名即可。
for j in range(1, 6):
    dataName = "data_batch_" + str(j)  # 读取当前目录下的data_batch12345文件，dataName其实也是data_batch文件的路径，本文和脚本文件在同一目录下。
    Xtr = unpickle(dataName)
    print dataName + " is loading..."

    for i in range(0, 10000):
        img = np.reshape(Xtr['data'][i], (3, 32, 32))  # Xtr['data']为图片二进制数据
        img = img.transpose(1, 2, 0)  # 读取image
        picName = 'train/' + str(Xtr['labels'][i]) + '_' + str(i + (j - 1)*10000) + '.jpg'  # Xtr['labels']为图片的标签，值范围0-9，本文中，train文件夹需要存在，并与脚本文件在同一目录下。
        imsave(picName, img)
    print dataName + " loaded."

print "test_batch is loading..."

# 生成测试集图片
testXtr = unpickle("test_batch")
for i in range(0, 10000):
    img = np.reshape(testXtr['data'][i], (3, 32, 32))
    img = img.transpose(1, 2, 0)
    picName = 'test/' + str(testXtr['labels'][i]) + '_' + str(i) + '.jpg'
    imsave(picName, img)
print "test_batch loaded."
```

**Comment by vieyahn2017** [comment]
## 机器学习零基础？手把手教你用TensorFlow搭建图像识别系统
https://blog.csdn.net/xiangz_csdn/article/details/54580637


CIFAR-10数据集。CIFAR-10包含了60000幅图片。它有10个不同的分类，每类包含6000幅图片。每幅图片的规格是32x32像素。这么小尺寸的图片对我们人类来说有时很难进行正确的分类，但它却简化了计算机模型的任务，并降低了分析图片的计算负载。


除了CIFAR-10，还有很多其他的图像数据集可以用于计算机视觉的研究。之所以使用标准数据集，有两个目的：

第一，要收集这么多的图片是一项巨大的工作。你需要有大量的图片资源，并根据需要对他们进行分别标签。

第二，使用相同的数据集，可以让我们有目的地比较不同方法的优劣。

另外，使用标准数据集，可以相互比较，在竞赛中排出名次。最有名的比赛是Image-Net图像识别大赛，它需要对1000个不同的种类进行识别。2012年的胜者是来自多伦多大学（University of Toronto）的亚力克斯•克利则夫斯基（Alex Krizhevsky），伊利亚·苏特斯科娃（Ilya Sutskever）和杰夫·辛顿（Geoffrey Hinton）设计的算法（雷锋网注：可点击论文链接查看）。这个系统领跑整个比赛并且以巨大的优势获胜。

这次比赛对整个研究领域产生了巨大的冲击，因为这是第一次使用卷积神经网络的方法获得胜利。卷积神经网络是一种人工神经网络，它大致模拟了动物视觉的皮质行为。这一技术的应用已经有一段时间，但它的潜力还没被多数人认识到。2012 Image-Net竞赛后，这种情况出现了改变。人们突然对神经网络和深度学习（深度学习是使用多层神经网络的方法解决机器学习的问题）产生了巨大兴趣，而这次赛事也极大地推动了以后几年深度学习的快速发展。
